### ISSUE TITLE
`bpkg`: avoid re-fetching the same repositories multiple times on the same fetch invocation

### ISSUE THREAD

**Issue by boris-kolpackov** [build2/build2#378]
From https://github.com/build2/build2/issues/167#issuecomment-1945759897:

> `bpkg` doesn't just fetch all the repositories and dumps the available packages in a big list. Rather, it creates sort of a tree of repositories listed as complement/prerequisite repositories and then it uses this information to see which packages/versions are "visible" to whom. The easiest way to build such a tree is by just fetching things recursively, which is I think what we are doing. An optimized version may see that we have already fetched some parts and avoid re-doing it.

Feels like there are two ways we could approach this: The first would be to simply copy the relevant available packages from our object model. If that doesn't fit for some reason (e.g., we don't store enough information), then we could create an in-memory cache of a repository and on subsequent encounters read this cache instead of re-fetching things over the network.

Also likely related to https://github.com/build2/build2/issues/350

**Comment by boris-kolpackov** [comment]
Ok, we've looked into this, both by reading the code and trying various tests, and the only obviously wrong behavior we've identified is the probing of remote `git` repository capabilities when all the required data has already been fetched. This has been fixed in https://github.com/build2/bpkg/commit/0d3aaa8f1ad1b6555189ba8381de3e4cb48fb40a and is available in the latest staged toolchain.

To elaborate on what works correctly:

1. For archive repositories we do not re-fetch the same repository multiple times within a single `bpkg-fetch` invocation.

2. For `git` repositories, we do not re-fetch commits that have already been fetched, nor call `ls-remote`, nor, with the above fix, check capabilities unless we need to do `ls-remote` or fetch some additional commits.

And so the summary here is that now, within a single `bpkg-fetch` invocation, we do things optimally.

Now that "within a single `bpkg-fetch` invocation" caveat is important: `bdep-init/bdep-sync` may perform multiple `bpkg-fetch` invocations under the hood if multiple linked configurations are involved (and `bdep-init` now forces a deep fetch; see #343 for background). So if multiple linked configurations are involved, then it may look like things are being re-fetched (we've now added some extra diagnostics to hopefully make it clear that fetching is happening in different configurations).

One idea to improve this "multiple `bpkg-fetch` calls under the hood" case would be to keep a cache of the fetched repository information on disk for the duration of, say, `bdep-init` under the assumption that withing the same `bdep-init` run it is unlikely to go stale. But this will be a "project", not a "quick fix". It's probably also makes sense to think about the local package cache (#167) since this looks related and will probably use the same underlying machanism/format.

@kamrann FYI

**Comment by kamrann** [comment]
Thanks for looking into this. I've just updated to latest staged and did a quick test. From the output it's unclear to me if what you're saying is holding true or not for my project, it's possible I'm misinterpreting the details/what exactly you're referring to by 'linked configurations'. 
[So here's the command output for reference](https://github.com/build2/build2/files/15478764/bdep_sync_fetch_output.txt).

Some things I've noted/am unsure about:

- Clearly there are multiple configurations active here. I do also have a couple of linked configurations that contain some third party packages (for example `base--opt`, which appears in the later `sync` output), but they are not showing up in the output from `bpkg-fetch`.
- There is only one `bpkg fetch` invocation listed, right at the top of the output.
- It seems there is always first a single chunk of `fetching...`/`querying...` lines, and then a series of similar chunks each preceded by `fetching in configuration...`. It's only in writing this that I'm starting to get a clearer idea of what's going on, but if I'm guessing correctly, the first is repository fetching (what packages and versions are available), whereas the rest is package fetching (fetching the contents of specific required package versions)?
- For archive packages, eg `cppget.org/stable`, there is a line saying it's being fetched in every configuration, all within seemingly a single `bpkg fetch` invocation; however it does appear that the real work is only being done once at the top, these subsequent lines of output do not appear to be associated with any significant time overhead.
- For git repositories it seems this is where it's problematic (though whether or not this differs from what you're saying should now be happening I'm not certain). `git:bitbucket.org/kamrann/kcore#main`, for example, is clearly being refetched multiple times. For reference, it was this package which I pushed a new commit to immediately before running the command.
- Even though the majority of packages had nothing new to actually pull, it feels like the bulk of the overhead is still coming just from all these repeated `querying`/`fetching` of git packages across configurations.
- I guess I don't really understand why the above command, invoked in a specific configuration, needed to fetch across multiple configurations in the first place? These are not configurations linked to each other, in the sense detailed in the docs; however they do each link to a common shared configuration. Is this the cause? From a purely practical point of view, the problem I'm having is that I have a bunch of projects initialized in a set of configurations, and even though at any one time I'm likely only actively building in one or two of those configurations, it's not uncommon for a `bdep update` command following some commits to trigger this kind of wide fetch across all configurations. I could of course try to deinit projects in configurations that I don't intend to work with for a while, but it's extra work/mental load, and it seems as if just having them remain initialized should not need to be adding overhead when updating other configurations.

If anything looks off to you and you need more info on my setup or want me to run any other commands, let me know.

**Comment by kamrann** [comment]
> Yes, `bdep-sync` synchronizes what we call a "configuration cluster", that is, a transitive set of all the linked configurations reachable from the configuration on which you are initiating the synchronization. This is the only way to get the transactional semantics.

Okay, so this is essentially what I was assuming, however I also assumed the links were one-directional and that this "reachability" would account for that. But from looking at the docs for `bpkg cfg-link`, the "implicit backlink" suggests that's maybe not the case?

Each of the configurations in your `grep` result was individually linked to a single `base--opt` configuration containing some third party packages; other than that no links were established, so they can all be thought of as sibling configurations.

In my head, the way this works is if I have configurations
  A
 /  \\
X   Y
where A contains some common dependencies, a `bdep update` in X would not need to fetch or sync with Y in any way, nor vice versa. If the links are in fact bidirectional, what are the transactional semantics that require this? Is it just that it implicitly supports looking for dependencies in both directions?

**Comment by boris-kolpackov** [comment]
> Each of the configurations in your grep result was individually linked to a single `base--opt` configuration containing some third party packages

Yes, I see it mentioned in your log but not with a `fetching in configuration` diagnostics line.

> If the links are in fact bidirectional, what are the transactional semantics that require this?

Continuing with your example, let's say you have changed a `depends` version constraint in a package in X for a dependency D that is in A. And Y also has a package that depends on D. Now if `bpkg` needs to upgrade/downgrade D in A in order to satisfy the new version constraint, we need to consider all the configurations that may have dependencies (and thus version constraints) on D in A.

**Comment by kamrann** [comment]
> Continuing with your example, let's say you have changed a `depends` version constraint in a package in X for a dependency D that is in A. And Y also has a package that depends on D. Now if `bpkg` needs to upgrade/downgrade D in A in order to satisfy the new version constraint, we need to consider all the configurations that may have dependencies (and thus version constraints) on D in A.

Okay makes sense. Though not entirely clear to me why "consider all the configurations" implies re-fetching them, as opposed to just taking into account their dependency requirements as they are now? If I follow right, you're trying to guarantee that in the following case:
```
bdep sync @X
[no changes to packages or configurations]
bdep sync @Y
```
that the sync of Y could never lead to any reconfiguration occurring? I guess I agree this behaviour is cleaner, but not immediately obvious why it would be important practically.

Anyway, since you say some things appear off in my output, perhaps best to figure that out. Any commands I can run or further info I can give that may shed some light?

**Comment by boris-kolpackov** [comment]
> Though not entirely clear to me why "consider all the configurations" implies re-fetching them, as opposed to just taking into account their dependency requirements as they are now?

This is what I meant by play the "let's hope we don't need to re-fetch this configuration this time" heuristics above. The problem with this approach is that in certain cases, without the up-to-date list of available packages in all the configurations in the cluster, we are guaranteed to make wrong decisions.

Again, continuing with the above example, let's say the change of the constraint in X cannot be reconciled with what we have in Y so we have to also upgrade/downgrade something in Y. Doing it without an up-to-date information of what's available is bound to produce bad results.

**Comment by boris-kolpackov** [comment]
> Any commands I can run or further info I can give that may shed some light?

@kamrann You've fixed a couple of bugs in the latest (today's) staged version -- could you upgrade and send the updated verbose output (it would be helpful if it were for exactly the same conditions as the previous log). Thanks!

**Comment by kamrann** [comment]
> Again, continuing with the above example, let's say the change of the constraint in X cannot be reconciled with what we have in Y so we have to also upgrade/downgrade something in Y. Doing it without an up-to-date information of what's available is bound to produce bad results.

Yep understood. I think my original view was can't we just kind of ignore Y when specifically syncing from X, and only worry about Y's view of things being satisfied after we next sync Y explicitly. But yeah more I think about it it doesn't make sense, you'd get ping-ponging with one config sync breaking another back and forth, and I guess in the end it would just defeat the point of configuration linking.

[Updated bdep sync -f output](https://github.com/user-attachments/files/15746548/bdep_sync_fetch_output_new.txt)

Okay, so the above output is from a close reproduction of the situation, not identical I'm afraid as my current multiple-project structure is a bit of a sprawling mess when it comes to git repos so resetting perfectly isn't simple. I now see multiple `bpkg fetch` invocations, and there are no `fetching in` occurrences - but I assume that is an intentional change to the output format. 

More to the point though, it looks good; there appear to be far fewer `git fetch` invocations, and nothing duplicated that I can see.  Does it look more like the output you'd expect? I'll continue to keep an eye on this going forwards, it's been a sporadic issue with occasional syncs taking way, way longer for reasons that weren't obvious to me, so can't say a lot from this single run. Looks promising though.

**Comment by boris-kolpackov** [comment]
Thanks, yes, this looks much better. We are now fairly certain the deep re-fetches in all the linked configurations were caused by a bug that was fixed in https://github.com/build2/bdep/commit/dfc89ec51c26d6226d767b79fe42e64af55e2b89

Thanks for your help in getting to the bottom of this!
